src/Microsoft.AspNet.Mvc.Core/Internal/Routing/LinkGenerationDecisionTree.cs

```diff
@@ -131,19 +131,38 @@ private class LinkGenerationMatchComparer : IComparer<LinkGenerationMatch>
 
             public int Compare(LinkGenerationMatch x, LinkGenerationMatch y)
             {
-                // For these comparisons lower is better.
-
+                // For this comparison lower is better.
                 if (x.Entry.Order != y.Entry.Order)
                 {
                     return x.Entry.Order.CompareTo(y.Entry.Order);
                 }
 
+                if (x.ParametersWithValues != y.ParametersWithValues)
+                {
+                    // If one match has more variables but they have the same segments we want the non-variable one
+                    // e.g. for 'help/general' and 'help/{id}' we want to make sure help/general is the first option
+                    if (x.Entry?.Template.Segments?.Count != y.Entry?.Template.Segments?.Count)
```

Comment:
If `Entry` is null we could have already thrown.

Did you actually find a case where any of these are null? I don't think `.Segments` can be null either.